From 810f3f1bef2f94f2c78945913576d365c4f09bb0 Mon Sep 17 00:00:00 2001
From: Manish Dharanenthiran <quic_mdharane@quicinc.com>
Date: Tue, 1 Nov 2022 09:24:23 +0530
Subject: [PATCH] ath12k: Modify ppdu info bandwidth in u_sig_cmn TLV

As part of EHT 2.0, channelization is supported in
HW. Because of this HW sends 320 BW update in two
different enums. Since, in monitor TLV update only
one enum is considered, BW update for stats update
is wrong which leads invalid rate calculations
in mac80211.

Modified the TLV update to use both 320 BW enum
before updating the rx stats.

Signed-off-by: Manish Dharanenthiran <quic_mdharane@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/dp_mon.c | 14 ++++++++++++--
 1 file changed, 12 insertions(+), 2 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/dp_mon.c b/drivers/net/wireless/ath/ath12k/dp_mon.c
index 1e21e92..0a3be9a 100644
--- a/drivers/net/wireless/ath/ath12k/dp_mon.c
+++ b/drivers/net/wireless/ath/ath12k/dp_mon.c
@@ -609,7 +609,16 @@ ath12k_dp_mon_hal_rx_parse_u_sig_cmn(u8 *rx_tlv, struct hal_rx_mon_ppdu_info *pp
 	ppdu_info->usig_common |= bad_usig_crc;
 	ppdu_info->u_sig_info.ul_dl = ul_dl;
 	ppdu_info->u_sig_info.bw = bandwidth;
-	ppdu_info->bw = bandwidth;
+
+	switch(bandwidth) {
+		case HAL_EHT_BW_320_2:
+		case HAL_EHT_BW_320_1:
+			ppdu_info->bw = ATH12K_BW_320;
+			break;
+		default:
+			ppdu_info->bw = bandwidth;
+			break;
+	}
 
 	return HAL_TLV_STATUS_PPDU_NOT_DONE;
 }
@@ -2158,7 +2167,7 @@ static void ath12k_dp_mon_rx_deliver_msdu(struct ath12k *ar, struct napi_struct
 	spin_unlock_bh(&ar->ab->base_lock);
 
 	ath12k_dbg(ar->ab, ATH12K_DBG_DATA,
-		   "rx skb %pK len %u peer %pM %u %s %s%s%s%s%s%s%s %srate_idx %u vht_nss %u freq %u band %u flag 0x%x fcs-err %i mic-err %i amsdu-more %i\n",
+		   "rx skb %pK len %u peer %pM %u %s %s%s%s%s%s%s%s%s %srate_idx %u vht_nss %u freq %u band %u flag 0x%x fcs-err %i mic-err %i amsdu-more %i\n",
 		   msdu,
 		   msdu->len,
 		   peer ? peer->addr : NULL,
@@ -2171,6 +2180,7 @@ static void ath12k_dp_mon_rx_deliver_msdu(struct ath12k *ar, struct napi_struct
 		   (status->bw == RATE_INFO_BW_40) ? "40" : "",
 		   (status->bw == RATE_INFO_BW_80) ? "80" : "",
 		   (status->bw == RATE_INFO_BW_160) ? "160" : "",
+		   (status->bw == RATE_INFO_BW_320) ? "320" : "",
 		   status->enc_flags & RX_ENC_FLAG_SHORT_GI ? "sgi " : "",
 		   status->rate_idx,
 		   status->nss,
-- 
2.17.1

